evm/vm/forks/sharding/shard_tracker.py

```diff
@@ -0,0 +1,161 @@
+        shard_id_topic_hex = encode_hex(self.shard_id.to_bytes(32, byteorder='big'))
+        new_logs = self.log_handler.get_new_logs(
+            address=self.vmc_address,
+            topics=[
+                encode_hex(self.COLLATION_ADDED_TOPIC),
+                shard_id_topic_hex,
+            ],
+        )
+        for log in new_logs:
+            yield parse_collation_added_log(log)
+
+    def get_next_log(self):
+        new_logs = self._get_new_logs()
+        self.new_logs.extend(new_logs)
+        if len(self.new_logs) == 0:
+            raise NextLogUnavailable("No more next logs")
+        return self.new_logs.pop()
+
+    # TODO: this method may return wrong result when new logs arrive before the logs inside
+    #       `self.new_logs` are consumed entirely. This issue can be resolved by saving the
+    #       status of `new_logs`, `unchecked_logs`, and `current_score`, when it start to run
+    #       `GUESS_HEAD`. If there is a new block arriving, just restore them to the saved status,
+    #       append new logs to `new_logs`, and re-run `GUESS_HEAD`
+    def fetch_candidate_head(self):
+        # Try to return a log that has the score that we are checking for,
+        # checking in order of oldest to most recent.
+        unchecked_logs = pipe(
+            self.unchecked_logs,
+            enumerate,
+            tuple,
+            reversed,
+            tuple,
+        )
+        for idx, log_entry in unchecked_logs:
+            if log_entry['score'] == self.current_score:
+                return self.unchecked_logs.pop(idx)
+        # If no further recorded but unchecked logs exist, go to the next
+        # is_new_head = true log
+        while True:
+            try:
+                log_entry = self.get_next_log()
+            # TODO: currently just raise when there is no log anymore
+            except NextLogUnavailable:
+                # TODO: should returns the genesis collation instead or just leave it?
+                raise NoCandidateHead("No candidate head available")
+            if log_entry['is_new_head']:
+                break
+            self.unchecked_logs.append(log_entry)
+        self.current_score = log_entry['score']
+        return log_entry
+
+    def clean_logs(self):
+        self.new_logs = []
+        self.unchecked_logs = []
+
+    def add_log(self, log_entry):
+        collation_height = log_entry['score']
+        while len(self.headers_by_height) <= collation_height:
+            self.headers_by_height.append([])
+        self.headers_by_height[collation_height].append(log_entry['header'])
+
+    def fetch_candidate_heads_generator(self, windback_length):
+        new_logs = self._get_new_logs()
+        for log_entry in new_logs:
+            # TODO: we can use another coroutine to sync the logs concurrently, instead of doing
+            #       that in this function
+            self.add_log(log_entry)
+        # TODO: deepcopy seems to be costly
```

Comment:
Considering this, it might be a bad idea to make `candidate_heads` a generator (whose state might change in between). As an alternative, we could make it a simple function and add a predicate argument. `candidate_head` would return the first head fulfilling the predicate.